compiler/src/iree/compiler/Codegen/Common/TileDispatchUsingForall.cpp

```diff
@@ -0,0 +1,202 @@
+  for (Operation *op : llvm::reverse(computeOps)) {
+    if (getLoweringConfig(op)) {
+      tilableOp = op;
+      break;
+    }
+  }
+  if (!tilableOp) {
+    // There is no lowering config. Return `null`.
+    return TilingInfo{nullptr, {}, {}};
+  }
+
+  IREE::Codegen::LoweringConfigAttrInterface tilableOpConfig =
+      getLoweringConfig(tilableOp);
+  if (!tilableOpConfig) {
+    return tilableOp->emitOpError("unable to find configuration of root op to "
+                                  "define workgroup count region");
+  }
+  auto tileSizes = llvm::map_to_vector(
+      tilableOpConfig.getWorkgroupTileSizes(),
+      [&](int64_t t) -> OpFoldResult { return rewriter.getIndexAttr(t); });
+  SmallVector<int64_t> interchange = tilableOpConfig.getWorkgroupInterchange();
+
+  // Avoid distributing unit-trip count loops.
+
+  // Set tile sizes for non-partitioned loops to zero.
+  if (auto partitionableLoopsInterface =
+          dyn_cast<PartitionableLoopsInterface>(tilableOp)) {
+    SmallVector<unsigned> partitionableLoops =
+        partitionableLoopsInterface.getPartitionableLoops(std::nullopt);
+    llvm::SmallDenseSet<unsigned> partitionableLoopsSet(
+        partitionableLoops.begin(), partitionableLoops.end());
+    OpFoldResult zero = rewriter.getIndexAttr(0);
+    for (auto loopId : llvm::seq<unsigned>(0, tileSizes.size())) {
+      if (partitionableLoopsSet.count(loopId)) {
+        continue;
+      }
+      tileSizes[loopId] = zero;
+    }
+  }
+
+  return TilingInfo{tilableOp, tileSizes, interchange};
+}
+
+/// Helper function to return the mapping attribute to use given the tile sizes.
+static SmallVector<Attribute> getMapping(MLIRContext *context,
+                                         ArrayRef<OpFoldResult> tileSizes) {
+  SmallVector<Attribute> mapping;
+  mapping.reserve(tileSizes.size());
+  for (auto tileSize : llvm::reverse(tileSizes)) {
+    if (isConstantIntValue(tileSize, 0)) {
+      continue;
+    }
+    uint64_t currSize = mapping.size();
+    switch (currSize) {
+    case 0:
+    case 1:
+    case 2:
+      mapping.push_back(IREE::Codegen::WorkgroupMappingAttr::get(
+          context, IREE::Codegen::symbolizeWorkgroupId(currSize).value()));
+      break;
+    default:
+      mapping.push_back(IREE::Codegen::WorkgroupMappingAttr::get(
+          context, IREE::Codegen::WorkgroupId::IdZ, currSize - 2));
+    }
+  }
+  return llvm::to_vector(llvm::reverse(mapping));
+}
+
+void TileAndDistributeToWorkgroupsUsingForallOpPass::runOnOperation() {
+  auto funcOp = getOperation();
+  auto *context = &getContext();
+  SmallVector<Operation *> computeOps = getComputeOps(funcOp);
+
+  IRRewriter rewriter(context);
+  FailureOr<TilingInfo> tilingInfo =
+      getTiledAndDistributionInfo(rewriter, computeOps);
+  if (failed(tilingInfo)) {
+    return signalPassFailure();
+  }
+  auto tilableOp = dyn_cast_or_null<TilingInterface>(tilingInfo->tilableOp);
+  if (!tilableOp) {
+    // Did not find a tileable op. So do nothing.
```

Comment:
Debug prints would be useful here.

Reply:
I typically dont rely on debug prints. So not sure what to put there. If you have suggestions, let me know (or if you send a PR, Ill stamp :P )